963 - Multi-Content Node: Adding page display type

shirleywbi opened the pull request:
## Changes
* Add Page style for MultiContent nodes.
    * Add Page component
    * Add SubPage component to handle "More Content". 
* Refactor AccordionMedia into MultiContentMedia, pulling out AccordionRows and PageRows as separate components
* Renaming several elements and functions to multi-content (from accordion)
* Bug Fix: Fix initial display when opening accordion and page LightBoxes
    * Added back missing activeIndex, setting accordion to show first child and page to show -1
* Bug Fix: Prevent autoplay when Multi-Content LightBox is opened for Youtube 
    * This is a work-around since `:autoplay=0` is not working.
* Bug Fix: Prevent autoplay when Multi-Content LightBox is opened for H5P Youtube
    * H5P youtube videos are based on Youtube Iframe API: https://developers.google.com/youtube/iframe_api_reference which is the same API used above.
    * I opted not to use setTimeout which could have also paused the video because there was a brief moment of sound that would play.
* Bug Fix: Console errors for when Multi-Content nodes have no children
* Add show title checkbox for page children

As of this PR, nested multi-content nodes appear with their parent node type (e.g., if I have a page MC, the sub-nodes will be page). This will be addressed in #971.

## Screenshot
![ezgif com-gif-maker (5)](https://user-images.githubusercontent.com/45089842/109378738-c8d22500-7889-11eb-99e6-5bbd30eab0cf.gif)
![image](https://user-images.githubusercontent.com/45089842/110707450-2598de00-81ae-11eb-93d7-40ef8a18579e.png)
![image](https://user-images.githubusercontent.com/45089842/110707515-45300680-81ae-11eb-989d-97b75f54898c.png)
## Issue Linkage
Closes #963 
## PR Dependency
Depends on: N/A
## Automated Testing
- Checks that text content should render
- Checks that given a text node with a more content node, both shall render
- Checks that given a video and if rows are locked until previous rows are completed, subsequent rows are locked

narendrasss commented:
@shirleywbi @wynnset Is it intentional that youtube content doesn't have titles?

<img width="888" alt="CleanShot 2021-03-03 at 11 45 56@2x" src="https://user-images.githubusercontent.com/31267630/109863060-08946600-7c16-11eb-85e4-104757b390e0.png">

shirleywbi commented:
> @shirleywbi @wynnset Is it intentional that youtube content doesn't have titles?
> 
> <img alt="CleanShot 2021-03-03 at 11 45 56@2x" width="888" src="https://user-images.githubusercontent.com/31267630/109863060-08946600-7c16-11eb-85e4-104757b390e0.png">

@wynnset Should I add a title to video content?

wynnset commented:
Is this an issue just for YouTube? It should be consistent across various content types. I think we should allow the user to define whether they want to show the title for each subsection in a multi-content node.

shirleywbi commented:
> Is this an issue just for YouTube? It should be consistent across various content types. I think we should allow the user to define whether they want to show the title for each subsection in a multi-content node.

@wynnset It's because it's showing tapestry-media. Videos in the lightbox don't usually show the title. It only shows the title in the node before clicking. I can add the title that the user defined for the node if you'd like.

narendrasss commented:
Found an issue: the content doesn't seem to show when you change a node to a multi content node.

![CleanShot 2021-03-30 at 13 03 27](https://user-images.githubusercontent.com/31267630/113049416-7a4dca00-9158-11eb-8b8b-21a5f42537fe.gif)

"Root" was initially a video node, and I changed it to multi content.

shirleywbi commented:
> Creating a new multi content node and adding child nodes work as expected, but a few issues:
> 
> 1. the content doesn't seem to show when you change a node to a multi content node.
> 
> ![CleanShot 2021-03-30 at 13 03 27](https://user-images.githubusercontent.com/31267630/113049416-7a4dca00-9158-11eb-8b8b-21a5f42537fe.gif)
> 
> "Root" was initially a video node, and I changed it to multi content.
> 
> 1. The show title checkbox isn't showing up for me:
> 
> <img alt="CleanShot 2021-03-30 at 13 08 50@2x" width="802" src="https://user-images.githubusercontent.com/31267630/113050021-37d8bd00-9159-11eb-8f30-43cb4b0f01f6.png">

@narendrasss Good call. I had the hardest time reproducing that first one since it only occurred when you had created non-multi-content nodes, refreshed, then changed one to multi-content since it dataset.js was hiding the childOrdering. The issues should be resolved now.

narendrasss reviewed (approved):
Looks good!

wynnset commented:
Locking mechanism is not working:
![image](https://user-images.githubusercontent.com/8981258/114779406-efa9c500-9d2a-11eb-859a-22f23803869e.png)
![image](https://user-images.githubusercontent.com/8981258/114779520-1536ce80-9d2b-11eb-9c77-05858546055d.png)

wynnset commented:
Locked node functionality is incomplete:
![image](https://user-images.githubusercontent.com/8981258/114780448-4532a180-9d2c-11eb-8fff-24075efc2dd3.png)
![image](https://user-images.githubusercontent.com/8981258/114780542-66938d80-9d2c-11eb-9826-77603121a4b5.png)

The locked icon is missing in the page menu. Also, if a node is locked, in the multi-content row for that node, we should show the following, centred, as the content:
- The lock icon
- "This content is current locked."
-  The conditions for its unlock (available currently when you hover over a locked node)  
![image](https://user-images.githubusercontent.com/8981258/114780402-3350fe80-9d2c-11eb-9119-b2c74e67d797.png)

shirleywbi commented:
@wynnset Added. Hope this is okay.
![image](https://user-images.githubusercontent.com/45089842/114797275-a87dfd00-9d47-11eb-86d6-f0092b18801f.png)

wynnset commented on templates/vue/src/components/Lightbox/media/MultiContentMedia/AccordionRows.vue:
```diff
@@ -0,0 +1,255 @@
+    },
+    rowBackground() {
+      if (this.isMultiContentContext) {
+        let rgb = 40
+        let colorOffset = this.level * 10
+        rgb = colorOffset > rgb ? 0 : rgb - colorOffset
+        return {
+          background: `rgb(${rgb}, ${rgb}, ${rgb})`,
+        }
```
Cool!

wynnset commented on templates/vue/src/components/Lightbox/media/H5PMedia/H5PIframe.vue:
```diff
@@ -349,11 +352,23 @@ export default {
+                  // This pause requires a timeout to avoid a spinner.
+                  setTimeout(() => {
+                    h5pVideo.pause()
+                  }, 1000)
```
Instead of this, could we do a `setInterval` and check to see if the video is loaded and if it is loaded, then pause the video and clear the interval? If it's not loaded (i.e. cannot be paused), we should keep going until the video is paused. The reason I am suggesting is that for me the video loads faster than 1000ms and therefore I actually hear the video playing for a bit under a second.

shirleywbi replied:
I tried using `setInterval` but it seems that by the time the H5P tells us the video has loaded as per loaded event found here: https://github.com/h5p/h5p-video, it may have already started playing. I opted to remove the `setTimeout` but there is a spinner.

wynnset reviewed (changes requested):
Just a few minor things:
- I updated the styling for a few things. I noticed you had the `<ul>` element repeating rather than `<li>` so I fixed that. I also simplified the styling syntax (I think there is some learning in there for you if you take a look at how I restructured that). Let me know if you notice any issues with any of that (I know the bullets got bigger but I like that look actually, and it’s ok that the locked menu items don’t get highlighted)
- [x] If I click to edit a child node directly in the tapestry, I expect that the close (X) button and cancel button would just close the modal, but they instead take me back up to the parent edit modal. Same thing if I save a child node. If a user is coming to the node modal from a parent node modal, we should keep track of that so we can take them back there. Otherwise, we should do the default behaviour of closing the modal.
- [x] Why is the styling for the title text of locked rows different than the ones not locked? Row titles should have consistent styling:  
![image](https://user-images.githubusercontent.com/8981258/115913786-e6ed7900-a425-11eb-83be-92c38b0e3526.png)

shirleywbi commented:
@wynnset 
> Just a few minor things:
> 
> * I updated the styling for a few things. I noticed you had the `<ul>` element repeating rather than `<li>` so I fixed that. I also simplified the styling syntax (I think there is some learning in there for you if you take a look at how I restructured that). Let me know if you notice any issues with any of that (I know the bullets got bigger but I like that look actually, and it’s ok that the locked menu items don’t get highlighted)

Thanks. This was really helpful.

> * [X]  If I click to edit a child node directly in the tapestry, I expect that the close (X) button and cancel button would just close the modal, but they instead take me back up to the parent edit modal. Same thing if I save a child node. If a user is coming to the node modal from a parent node modal, we should keep track of that so we can take them back there. Otherwise, we should do the default behaviour of closing the modal.
> * [X]  Why is the styling for the title text of locked rows different than the ones not locked? Row titles should have consistent styling

Fixed. The styling differed because the title was being generated differently for locked and unlocked nodes. When locked, the title is generated by PageRows. When unlocked, the title is from the given TapestryMedia.

wynnset reviewed (changes requested):
This is pretty much good to go now. I've gone through and updated the code and styling where needed myself except for a couple of things:
- I find the language "isSubMultiContentRow", "hasSubMultiContent", or "subMultiContent" really hard to understand... any way we can change that to something easier to grasp?
- I noticed that when I log in as a user with edit access to a multi-content node, I do not see the multi-content node's child nodes (I should). If I log in as the user who created the node, I can see those nodes.

Once these are addressed, please mark as "ready to merge". Thanks @shirleywbi !

shirleywbi commented:
> * I noticed that when I log in as a user with edit access to a multi-content node, I do not see the multi-content node's child nodes (I should). If I log in as the user who created the node, I can see those nodes.
> 
@wynnset I can't seem to replicate this. For example, I created the nodes using an admin account. I'm getting all my nodes as an editor and author.
![image](https://user-images.githubusercontent.com/45089842/116113476-60c37380-a66d-11eb-979b-b48ec085441c.png)
![image](https://user-images.githubusercontent.com/45089842/116113584-75a00700-a66d-11eb-9d70-29517c42cef4.png)

wynnset commented:
@shirleywbi sorry never mind. I had only set the main node to be visible and not the subnodes, so it's ok that it didn't show up. Ignore that part of my comments!

shirleywbi commented:
@wynnset Since I made a couple of extra changes, I marked this as "ready for review".
